### ISSUE TITLE
Search by chemical timeout

### ISSUE THREAD

**Issue by thomasoide** [rji-futures-lab/django-rmp-data#101]
The search by chemical function is timing out for some chemicals. Timeout happens when processing larger querysets.

**Comment by gordonje** [comment]
I get the same timeout when the user selects the first option "Public OCA Chemical".

**Comment by thomasoide** [comment]
The fact that query is so slow makes sense. Based on the last time I looked at the data, it looked like every facility in the dataset had at least one Public OCA Chemical.

**Comment by gordonje** [comment]
Tried updating from the RDS instance from t2.micro to t2.medium. That's doubling the vCPU and quadrupling the memory. But I'm still getting the API Gateway timeout.

Ran the same search while serving locally (so with even more CPU and RAM available to postgres), and I clocked it at about a minute and half.

My guess either or both of following would help:

1. **Pagination:** This particular option returns 11,916 distinct facilities, and we are currently loading them all on the page. Currently, we aren't making use of Django [Paginator](https://docs.djangoproject.com/en/2.2/topics/pagination/) class in any of our ListView. Seems like should. Although, now that the js tables are working again, I also see a "Show ## entries" option that doesn't seem to be working? Maybe we need to do something else to get [pagination working with datatables.net](https://datatables.net/reference/option/paging)?

2. **Re-write the Django QuerySet:** This is just a hunch on my part, I think there might be a more optimal way to get distinct facilities with the quantity (in pounds) of the selected chemical. I'm going to take a closer look at the SQL...

**Comment by gordonje** [comment]
Here is the current SQL:

```sql
SELECT DISTINCT ON ("rmp_process"."facility_id") 
    "rmp_procchem"."procchem_id", "rmp_procchem"."process_id", 
    "rmp_procchem"."chemical_id", "rmp_procchem"."quantity_lbs",
    "rmp_procchem"."cbi_flag", "rmp_procchem"."num_alt_flam",
    "rmp_procchem"."num_alt_tox", "rmp_procchem"."num_prevent_2_chem",
    "rmp_procchem"."num_prevent_3_chem", "rmp_procchem"."num_proc_flam",
    "rmp_procchem"."num_worst_flam", "rmp_procchem"."num_worst_tox",
    "rmp_procchem"."cas", "rmp_procchem"."chemical_type",
    "rmp_procchem"."chemical_name", "rmp_procchem"."worst_tox_flag",
    "rmp_procchem"."worst_flam_flag", "rmp_process"."id",
    "rmp_process"."process_desc", "rmp_process"."rmp_id",
    "rmp_process"."program_level", "rmp_process"."cbi_flag",
    "rmp_process"."num_proc_chem", "rmp_process"."num_proc_naics",
    "rmp_process"."num_worst_tox", "rmp_process"."num_worst_flam",
    "rmp_process"."num_alt_tox", "rmp_process"."num_alt_flam",
    "rmp_process"."toxic_tot", "rmp_process"."flam_tot",
    "rmp_process"."quantity_tot", "rmp_process"."facility_name",
    "rmp_process"."facility_id", "rmp_process"."rmp_receipt_date"
FROM "rmp_procchem" 
INNER JOIN "rmp_process"
ON ("rmp_procchem"."process_id" = "rmp_process"."id")
WHERE 
    to_tsvector(COALESCE("rmp_procchem"."chemical_name", )) @@ (plainto_tsquery(Public OCA Chemical)) = true
ORDER BY "rmp_process"."facility_id" ASC, "rmp_process"."rmp_receipt_date" DESC;
```

**Comment by gordonje** [comment]
@thomasoide There might even be a fallback hack solution for this particular search option: Return every facility with a 0 quantity.

Still, I'm fiddling around with the queryset to see if there's another way.

**Comment by thomasoide** [comment]
I just went through the site and it appears that the datatables are working on all pages. The paginate by # feature is disabled on 5-year accident histories and accidents by state because i overrode the defaults to show all of the results on the page. 

That being said, you're right on the Public OCA stuff. I played around with the paginate_by attribute on the ListView and it was super buggy on the front end, so it probably is worth going back in and giving it another look. Using datatables was the quick fix

**Comment by thomasoide** [comment]
@gordonje We would have to write a special view for that particular case if we did do that, correct?

**Comment by mizzousundevil** [comment]
This doesn't solve the whole problem, but the Public OCA search in the chemical list would be pretty useless. Every facility would have one.

**Comment by thomasoide** [comment]
That is true, we could just decide to omit it altogether. I don't think any other queryset is as large, so it could work.

**Comment by gordonje** [comment]
Sorry I checked out for a bit.

I've excluded the Public OCA Chemical option, and deployed that change. Still do some extra work around the edges of the search queries. Will keep reporting changes in this thread.

**Comment by gordonje** [comment]
@mizzousundevil I just pushed up some changes that improve the performance of the search by chemical queries. The results are slightly different than before, but I think closer to what folks expect (e.g., limiting to only facilities with a *current* rmp with processes that include those chemicals).

Check it (or have others do so) and let us know.

**Comment by mizzousundevil** [comment]
Still times out for Anhydrous Ammonia

**Comment by mizzousundevil** [comment]
That's the only chemical I know that's timing out right now based on searching random common chems

**Comment by mizzousundevil** [comment]
Still getting a timeout on the Ammonia (anhydrous) search

**Comment by gordonje** [comment]
@mizzousundevil Thanks for catching this.

I just pushed another tweak that cuts query time in half. But I'm still getting the time out. I think the fundamental issue is the number of records we are rendering on the page (over 10k facilities currently have processes that include Ammonia). The way the page currently works is that we read all those rows from the db, then use the js plug into paginate them. That makes initial load slow, but moving between the pages super-fast.

We might try looking at integrating Django's built in paginator. Maybe only when the count of results is super-high.

Keeping this issue open...
